File: mobile/android/app/src/main/java/co/hndrx/hearth/core/gateway/GatewayClient.kt
package co.hndrx.hearth.core.gateway

import co.hndrx.hearth.core.auth.AuthManager
import co.hndrx.hearth.core.models.ConnectionState
import co.hndrx.hearth.core.models.GatewayEvent
import co.hndrx.hearth.core.models.GatewayMessage
import co.hndrx.hearth.core.models.GatewayOpcode
import com.squareup.moshi.Moshi
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import javax.inject.Inject
import javax.inject.Singleton
import kotlin.math.min
import kotlin.random.Random

@Singleton
class GatewayClient @Inject constructor(
    private val okHttpClient: OkHttpClient,
    private val moshi: Moshi,
    private val authManager: AuthManager,
) {

    companion object {
        private const val MAX_RECONNECT_ATTEMPTS = 10
        private const val BASE_RECONNECT_DELAY_MS = 1000L
        private const val MAX_RECONNECT_DELAY_MS = 60_000L
    }

    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())
    private val gatewayMessageAdapter = moshi.adapter(GatewayMessage::class.java)

    private var webSocket: WebSocket? = null
    private var heartbeatJob: Job? = null
    private var heartbeatIntervalMs: Long = 0L
    private var lastHeartbeatAcked: Boolean = true
    private var sequence: Int? = null
    private var sessionId: String? = null
    private var reconnectAttempts: Int = 0
    private var gatewayUrl: String = ""

    private val _connectionState = MutableStateFlow(ConnectionState.DISCONNECTED)
    val connectionState: StateFlow<ConnectionState> = _connectionState.asStateFlow()

    private val _events = MutableSharedFlow<GatewayEvent>(extraBufferCapacity = 256)
    val events: SharedFlow<GatewayEvent> = _events.asSharedFlow()

    fun connect(url: String) {
        gatewayUrl = url
        reconnectAttempts = 0
        openWebSocket()
    }

    fun disconnect() {
        heartbeatJob?.cancel()
        heartbeatJob = null
        webSocket?.close(1000, "Client disconnect")
        webSocket = null
        _connectionState.value = ConnectionState.DISCONNECTED
        sessionId = null
        sequence = null
    }

    private fun openWebSocket() {
        _connectionState.value = ConnectionState.CONNECTING

        val token = authManager.getAccessToken() ?: return
        val wsUrl = if (gatewayUrl.contains("?")) {
            "$gatewayUrl&token=$token"
        } else {
            "$gatewayUrl?token=$token"
        }

        val request = Request.Builder()
            .url(wsUrl)
            .build()

        webSocket = okHttpClient.newWebSocket(request, object : WebSocketListener() {
            override fun onOpen(webSocket: WebSocket, response: Response) {
                _connectionState.value = ConnectionState.CONNECTED
                reconnectAttempts = 0
            }

            override fun onMessage(webSocket: WebSocket, text: String) {
                handleMessage(text)
            }

            override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
                webSocket.close(code, reason)
            }

            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                heartbeatJob?.cancel()
                if (_connectionState.value != ConnectionState.DISCONNECTED) {
                    attemptReconnect()
                }
            }

            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                heartbeatJob?.cancel()
                if (_connectionState.value != ConnectionState.DISCONNECTED) {
                    attemptReconnect()
                }
            }
        })
    }

    private fun handleMessage(text: String) {
        val message = try {
            gatewayMessageAdapter.fromJson(text) ?: return
        } catch (_: Exception) {
            return
        }

        // Track sequence number
        message.s?.let { sequence = it }

        when (message.op) {
            GatewayOpcode.HELLO -> {
                handleHello(message)
            }
            GatewayOpcode.HEARTBEAT -> {
                sendHeartbeat()
            }
            GatewayOpcode.HEARTBEAT_ACK -> {
                lastHeartbeatAcked = true
            }
            GatewayOpcode.DISPATCH -> {
                handleDispatch(message)
            }
            GatewayOpcode.RECONNECT -> {
                webSocket?.close(4000, "Server requested reconnect")
                attemptReconnect()
            }
            GatewayOpcode.INVALID_SESSION -> {
                val resumable = message.d as? Boolean ?: false
                if (!resumable) {
                    sessionId = null
                    sequence = null
                }
                scope.launch {
                    delay(Random.nextLong(1000, 5000))
                    openWebSocket()
                }
            }
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun handleHello(message: GatewayMessage) {
        val data = message.d as? Map<String, Any?> ?: return
        val interval = (data["heartbeat_interval"] as? Number)?.toLong() ?: 41250L
        heartbeatIntervalMs = interval
        startHeartbeat()

        // Send IDENTIFY or RESUME
        if (sessionId != null && sequence != null) {
            sendResume()
        } else {
            sendIdentify()
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun handleDispatch(message: GatewayMessage) {
        val eventType = message.t ?: return

        // Capture session ID from READY event
        if (eventType == "READY") {
            val data = message.d as? Map<String, Any?> ?: return
            sessionId = data["session_id"] as? String
        }

        scope.launch {
            _events.emit(
                GatewayEvent(
                    type = eventType,
                    data = message.d,
                    sequence = message.s,
                )
            )
        }
    }

    private fun startHeartbeat() {
        heartbeatJob?.cancel()
        lastHeartbeatAcked = true
        heartbeatJob = scope.launch {
            // Initial jitter delay
            delay(Random.nextLong(0, heartbeatIntervalMs))
            while (true) {
                if (!lastHeartbeatAcked) {
                    // Missed heartbeat ACK -- zombie connection
                    webSocket?.close(4009, "Heartbeat ACK missed")
                    attemptReconnect()
                    return@launch
                }
                sendHeartbeat()
                delay(heartbeatIntervalMs)
            }
        }
    }

    private fun sendHeartbeat() {
        lastHeartbeatAcked = false
        val payload = GatewayMessage(op = GatewayOpcode.HEARTBEAT, d = sequence)
        send(payload)
    }

    private fun sendIdentify() {
        val token = authManager.getAccessToken() ?: return
        val identifyData = mapOf(
            "token" to token,
            "properties" to mapOf(
                "os" to "Android",
                "browser" to "Hearth Android",
                "device" to "Hearth Android",
            ),
        )
        val payload = GatewayMessage(op = GatewayOpcode.IDENTIFY, d = identifyData)
        send(payload)
    }

    private fun sendResume() {
        val token = authManager.getAccessToken() ?: return
        val resumeData = mapOf(
            "token" to token,
            "session_id" to sessionId,
            "seq" to sequence,
        )
        val payload = GatewayMessage(op = GatewayOpcode.RESUME, d = resumeData)
        send(payload)
    }

    fun sendPresenceUpdate(status: String) {
        val presenceData = mapOf(
            "status" to status,
            "afk" to false,
        )
        val payload = GatewayMessage(op = GatewayOpcode.PRESENCE_UPDATE, d = presenceData)
        send(payload)
    }

    private fun send(message: GatewayMessage) {
        val json = try {
            gatewayMessageAdapter.toJson(message)
        } catch (_: Exception) {
            return
        }
        webSocket?.send(json)
    }

    private fun attemptReconnect() {
        if (reconnectAttempts >= MAX_RECONNECT_ATTEMPTS) {
            _connectionState.value = ConnectionState.DISCONNECTED
            return
        }

        _connectionState.value = ConnectionState.RECONNECTING
        reconnectAttempts++

        scope.launch {
            val baseDelay = BASE_RECONNECT_DELAY_MS * (1L shl min(reconnectAttempts - 1, 10))
            val cappedDelay = min(baseDelay, MAX_RECONNECT_DELAY_MS)
            val jitter = Random.nextLong(0, cappedDelay / 2)
            delay(cappedDelay + jitter)
            openWebSocket()
        }
    }
}

File: mobile/ios/Hearth/Core/Gateway/GatewayClient.swift
import Foundation

/// Events emitted by the gateway client to consumers.
enum GatewayEvent: Sendable {
    case connected
    case disconnected(Error?)
    case dispatch(eventName: String, data: [String: Any])
    case invalidSession(resumable: Bool)
}

actor GatewayClient {
    private let baseURL: URL
    private var webSocketTask: URLSessionWebSocketTask?
    private let session: URLSession

    private var token: String?
    private var sessionId: String?
    private var sequence: Int?
    private var heartbeatInterval: TimeInterval = 41.25
    private var heartbeatTask: Task<Void, Never>?
    private var receiveTask: Task<Void, Never>?
    private var lastHeartbeatAck = true
    private var reconnectAttempts = 0
    private let maxReconnectAttempts = 10
    private var isConnected = false

    private var eventContinuation: AsyncStream<GatewayEvent>.Continuation?

    /// An async stream of gateway events for consumers to observe.
    let events: AsyncStream<GatewayEvent>

    init(baseURL: URL, session: URLSession = .shared) {
        self.baseURL = baseURL
        self.session = session

        var continuation: AsyncStream<GatewayEvent>.Continuation!
        self.events = AsyncStream { continuation = $0 }
        self.eventContinuation = continuation
    }

    deinit {
        eventContinuation?.finish()
    }

    // MARK: - Public API

    func connect(token: String) {
        self.token = token
        reconnectAttempts = 0
        establishConnection()
    }

    func disconnect() {
        heartbeatTask?.cancel()
        heartbeatTask = nil
        receiveTask?.cancel()
        receiveTask = nil
        webSocketTask?.cancel(with: .goingAway, reason: nil)
        webSocketTask = nil
        isConnected = false
        eventContinuation?.yield(.disconnected(nil))
    }

    func send<T: Encodable>(op: GatewayOpcode, data: T) async throws {
        let payload = GatewayOutbound(op: op.rawValue, d: data)
        let jsonData = try JSONEncoder().encode(payload)
        guard let string = String(data: jsonData, encoding: .utf8) else { return }
        try await webSocketTask?.send(.string(string))
    }

    func updatePresence(status: String, activities: [[String: Any]]? = nil) async throws {
        var payload: [String: Any] = ["status": status]
        if let activities {
            payload["activities"] = activities
        }
        let data = try JSONSerialization.data(withJSONObject: payload)
        let message = GatewayOutbound(
            op: GatewayOpcode.presenceUpdate.rawValue,
            d: AnyCodable(try JSONSerialization.jsonObject(with: data))
        )
        let encoded = try JSONEncoder().encode(message)
        guard let string = String(data: encoded, encoding: .utf8) else { return }
        try await webSocketTask?.send(.string(string))
    }

    var currentSequence: Int? {
        sequence
    }

    var currentSessionId: String? {
        sessionId
    }

    // MARK: - Connection Management

    private func establishConnection() {
        var urlComponents = URLComponents(url: baseURL.appendingPathComponent("gateway"), resolvingAgainstBaseURL: true)
        if let token {
            urlComponents?.queryItems = [URLQueryItem(name: "token", value: token)]
        }

        guard let url = urlComponents?.url else { return }

        let task = session.webSocketTask(with: url)
        self.webSocketTask = task
        task.resume()

        receiveTask?.cancel()
        receiveTask = Task { [weak self] in
            await self?.receiveLoop()
        }
    }

    private func receiveLoop() async {
        guard let task = webSocketTask else { return }

        while !Task.isCancelled {
            do {
                let message = try await task.receive()
                switch message {
                case .string(let text):
                    if let data = text.data(using: .utf8) {
                        await handleMessage(data)
                    }
                case .data(let data):
                    await handleMessage(data)
                @unknown default:
                    break
                }
            } catch {
                if !Task.isCancelled {
                    await handleDisconnect(error: error)
                }
                return
            }
        }
    }

    // MARK: - Message Handling

    private func handleMessage(_ data: Data) async {
        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let opRaw = json["op"] as? Int,
              let op = GatewayOpcode(rawValue: opRaw) else {
            return
        }

        let d = json["d"]
        let s = json["s"] as? Int
        let t = json["t"] as? String

        if let s {
            sequence = s
        }

        switch op {
        case .hello:
            await handleHello(d: d)

        case .heartbeatAck:
            lastHeartbeatAck = true

        case .dispatch:
            await handleDispatch(eventName: t, data: d)

        case .heartbeat:
            await sendHeartbeat()

        case .reconnect:
            await attemptReconnect()

        case .invalidSession:
            let resumable = (d as? Bool) ?? false
            if !resumable {
                sessionId = nil
                sequence = nil
            }
            eventContinuation?.yield(.invalidSession(resumable: resumable))
            if resumable {
                try? await Task.sleep(nanoseconds: UInt64.random(in: 1_000_000_000...5_000_000_000))
                await attemptReconnect()
            } else {
                try? await Task.sleep(nanoseconds: UInt64.random(in: 1_000_000_000...5_000_000_000))
                establishConnection()
            }

        default:
            break
        }
    }

    private func handleHello(d: Any?) async {
        if let dict = d as? [String: Any],
           let interval = dict["heartbeat_interval"] as? Int {
            heartbeatInterval = Double(interval) / 1000.0
        }

        isConnected = true
        reconnectAttempts = 0
        lastHeartbeatAck = true
        eventContinuation?.yield(.connected)

        startHeartbeat()

        // Send IDENTIFY or RESUME
        if let sessionId, let sequence {
            await sendResume(sessionId: sessionId, sequence: sequence)
        } else {
            await sendIdentify()
        }
    }

    private func handleDispatch(eventName: String?, data: Any?) async {
        guard let eventName else { return }

        // Capture session_id from READY event
        if eventName == "READY", let dict = data as? [String: Any] {
            sessionId = dict["session_id"] as? String
        }

        let eventData = (data as? [String: Any]) ?? [:]
        eventContinuation?.yield(.dispatch(eventName: eventName, data: eventData))
    }

    private func handleDisconnect(error: Error?) async {
        isConnected = false
        heartbeatTask?.cancel()
        heartbeatTask = nil
        eventContinuation?.yield(.disconnected(error))
        await attemptReconnect()
    }

    // MARK: - Heartbeat

    private func startHeartbeat() {
        heartbeatTask?.cancel()
        heartbeatTask = Task { [weak self] in
            guard let self else { return }

            // Initial jitter: wait a random fraction of the heartbeat interval before the first heartbeat
            let jitterNanos = UInt64(Double.random(in: 0...1) * await self.heartbeatInterval * 1_000_000_000)
            try? await Task.sleep(nanoseconds: jitterNanos)

            while !Task.isCancelled {
                let interval = await self.heartbeatInterval
                let ackReceived = await self.lastHeartbeatAck

                if !ackReceived {
                    // Missed heartbeat ACK - reconnect
                    await self.attemptReconnect()
                    return
                }

                await self.sendHeartbeat()
                try? await Task.sleep(nanoseconds: UInt64(interval * 1_000_000_000))
            }
        }
    }

    private func sendHeartbeat() async {
        lastHeartbeatAck = false
        let payload: [String: Any?] = ["op": GatewayOpcode.heartbeat.rawValue, "d": sequence]
        guard let data = try? JSONSerialization.data(withJSONObject: payload.compactMapValues { $0 ?? NSNull() }),
              let string = String(data: data, encoding: .utf8) else {
            return
        }
        try? await webSocketTask?.send(.string(string))
    }

    // MARK: - Identify / Resume

    private func sendIdentify() async {
        guard let token else { return }

        let identify = GatewayIdentify(
            token: token,
            properties: GatewayIdentify.IdentifyProperties(
                os: "iOS",
                browser: "Hearth iOS",
                device: "iPhone"
            )
        )

        try? await send(op: .identify, data: identify)
    }

    private func sendResume(sessionId: String, sequence: Int) async {
        guard let token else { return }
        let resume = GatewayResume(token: token, sessionId: sessionId, seq: sequence)
        try? await send(op: .resume, data: resume)
    }

    // MARK: - Reconnect

    private func attemptReconnect() async {
        guard reconnectAttempts < maxReconnectAttempts else {
            eventContinuation?.yield(.disconnected(
                NSError(domain: "GatewayClient", code: -1,
                        userInfo: [NSLocalizedDescriptionKey: "Max reconnect attempts reached"])
            ))
            return
        }

        webSocketTask?.cancel(with: .goingAway, reason: nil)
        webSocketTask = nil
        heartbeatTask?.cancel()
        heartbeatTask = nil

        reconnectAttempts += 1

        // Exponential backoff with jitter: base * 2^attempt + random jitter
        let base: Double = 1.0
        let backoff = base * pow(2.0, Double(min(reconnectAttempts, 6)))
        let jitter = Double.random(in: 0...backoff * 0.5)
        let delay = backoff + jitter

        try? await Task.sleep(nanoseconds: UInt64(delay * 1_000_000_000))

        guard !Task.isCancelled else { return }
        establishConnection()
    }
}

// MARK: - Outbound payload helper

private struct GatewayOutbound<T: Encodable>: Encodable {
    let op: Int
    let d: T
}
